crates/transpiler/src/passes/disjoint_layout.rs

```diff
@@ -0,0 +1,545 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use hashbrown::{HashMap, HashSet};
+
+use pyo3::create_exception;
+use pyo3::intern;
+use pyo3::prelude::*;
+use pyo3::types::{PyList, PyTuple};
+use rayon::prelude::*;
+use rustworkx_core::connectivity::connected_components;
+use rustworkx_core::petgraph::prelude::*;
+use rustworkx_core::petgraph::visit::{IntoEdgeReferences, IntoNodeReferences, NodeFiltered};
+use rustworkx_core::petgraph::EdgeType;
+use smallvec::SmallVec;
+use uuid::Uuid;
+
+use qiskit_circuit::bit::ShareableQubit;
+use qiskit_circuit::converters::circuit_to_dag;
+use qiskit_circuit::dag_circuit::DAGCircuit;
+use qiskit_circuit::imports::ImportOnceCell;
+use qiskit_circuit::operations::{Operation, OperationRef, Param, StandardInstruction};
+use qiskit_circuit::packed_instruction::PackedOperation;
+use qiskit_circuit::{Clbit, PhysicalQubit, Qubit, VirtualQubit};
+
+use crate::target::{Qargs, Target};
+use crate::TranspilerError;
+
+create_exception!(qiskit, MultiQEncountered, pyo3::exceptions::PyException);
+
+static COUPLING_MAP: ImportOnceCell =
+    ImportOnceCell::new("qiskit.transpiler.coupling", "CouplingMap");
+
+pub struct DisjointComponent {
+    pub physical_qubits: Vec<PhysicalQubit>,
+    pub sub_dag: DAGCircuit,
+    pub virtual_qubits: Vec<VirtualQubit>,
+}
+
+pub enum DisjointSplit {
+    NoneNeeded,
+    TargetSubset(Vec<PhysicalQubit>),
+    Arbitrary(Vec<DisjointComponent>),
+}
+
+type CouplingMap = StableDiGraph<NodeIndex, ()>;
```

Comment:
1. does this need to be a `StableGraph` over a `Graph`?  Also, is this graph _definitely_ directed?  Given that we've erased the basis-gates information from the `Target`, I'd argue that it's meaningless to try and call it directed and we should make it undirected.
2. which graph is the `NodeIndex` a reference into?  I _think_ eventually they refer to `PhysicalQubit`s in the `Target`?  Maybe we can just make the weights `PhysicalQubit`?

Reply:
I think I can probably get away with a `Graph` there shouldn't be any removals, but also the canonical index is the node weight which represents the `PhysicalQubit` (which I'll update).  I also think we can move it to be undirected, since in current state of the PR we're only getting a physical qubit list and not passing around a graph object directly.

Reply:
I updated this in: https://github.com/Qiskit/qiskit/pull/14177/commits/cd7ad0a72033eaa5b9f3e3746fda77e02480e0ef there is one small nuance that `CouplingMap` in python is explicitly directed, but it's simple enough to workaround this. It only effects dense layout since sabre always works with undirected graphs.